File: backend/routes/lawyers.py
from flask import Blueprint, request, jsonify
import json
from database.db import get_db
from middleware.auth import authenticate_token, require_admin

lawyers_bp = Blueprint('lawyers', __name__)

@lawyers_bp.route('/', methods=['GET'])
def get_lawyers():
    try:
        practice_area = request.args.get('practiceArea')
        state = request.args.get('state')
        min_experience = request.args.get('minExperience', type=int)
        max_rate = request.args.get('maxRate', type=float)
        response_guarantee = request.args.get('responseGuarantee') == 'true'
        sort_by = request.args.get('sortBy', 'id')
        
        conn = get_db()
        cursor = conn.cursor()
        
        query = 'SELECT * FROM lawyers WHERE 1=1'
        params = []
        
        if practice_area:
            query += ' AND practice_area = ?'
            params.append(practice_area)
        
        if state:
            query += ' AND location_state = ?'
            params.append(state)
        
        if min_experience:
            query += ' AND experience_years >= ?'
            params.append(min_experience)
        
        if max_rate:
            query += ' AND hourly_rate_min <= ?'
            params.append(max_rate)
        
        if response_guarantee:
            query += ' AND response_guarantee = 1'
        
        # Sorting
        valid_sorts = {'id': 'id', 'experience_years': 'experience_years', 
                      'hourly_rate_min': 'hourly_rate_min', 'success_rate': 'success_rate'}
        sort_field = valid_sorts.get(sort_by, 'id')
        query += f' ORDER BY {sort_field} DESC'
        
        cursor.execute(query, params)
        lawyers = [dict(row) for row in cursor.fetchall()]
        conn.close()
        
        # Format lawyers
        formatted_lawyers = []
        for lawyer in lawyers:
            formatted_lawyer = {
                **lawyer,
                'specialties': json.loads(lawyer['specialties']) if lawyer['specialties'] else [],
                'verified': bool(lawyer['verified']),
                'mediationCertified': bool(lawyer['mediation_certified']),
                'responseGuarantee': bool(lawyer['response_guarantee']),
                'practiceArea': lawyer['practice_area'],
                'experienceYears': lawyer['experience_years'],
                'caseCount': lawyer['case_count'],
                'successRate': lawyer['success_rate'],
                'hourlyRateMin': lawyer['hourly_rate_min'],
                'hourlyRateMax': lawyer['hourly_rate_max'],
                'locationCity': lawyer['location_city'],
                'locationState': lawyer['location_state'],
                'maraNumber': lawyer['mara_number'],
                'avatarColor': lawyer['avatar_color']
            }
            formatted_lawyers.append(formatted_lawyer)
        
        return jsonify({'lawyers': formatted_lawyers})
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@lawyers_bp.route('/<int:lawyer_id>', methods=['GET'])
def get_lawyer(lawyer_id):
    try:
        conn = get_db()
        cursor = conn.cursor()
        cursor.execute('SELECT * FROM lawyers WHERE id = ?', (lawyer_id,))
        lawyer_row = cursor.fetchone()
        conn.close()
        
        if not lawyer_row:
            return jsonify({'error': 'Lawyer not found'}), 404
        
        lawyer = dict(lawyer_row)
        formatted_lawyer = {
            **lawyer,
            'specialties': json.loads(lawyer['specialties']) if lawyer['specialties'] else [],
            'verified': bool(lawyer['verified']),
            'mediationCertified': bool(lawyer['mediation_certified']),
            'responseGuarantee': bool(lawyer['response_guarantee']),
            'practiceArea': lawyer['practice_area'],
            'experienceYears': lawyer['experience_years'],
            'caseCount': lawyer['case_count'],
            'successRate': lawyer['success_rate'],
            'hourlyRateMin': lawyer['hourly_rate_min'],
            'hourlyRateMax': lawyer['hourly_rate_max'],
            'locationCity': lawyer['location_city'],
            'locationState': lawyer['location_state'],
            'maraNumber': lawyer['mara_number'],
            'avatarColor': lawyer['avatar_color']
        }
        
        return jsonify({'lawyer': formatted_lawyer})
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@lawyers_bp.route('/', methods=['POST'])
@authenticate_token
@require_admin
def create_lawyer():
    try:
        data = request.get_json()
        
        required_fields = ['name', 'firm', 'practiceArea', 'experienceYears', 
                          'hourlyRateMin', 'hourlyRateMax', 'locationCity', 'locationState']
        if not all(data.get(field) for field in required_fields):
            return jsonify({'error': 'Missing required fields'}), 400
        
        conn = get_db()
        cursor = conn.cursor()
        
        cursor.execute('''
            INSERT INTO lawyers (
                name, firm, tier, practice_area, specialties, experience_years,
                case_count, success_rate, hourly_rate_min, hourly_rate_max,
                location_city, location_state, verified, mediation_certified,
                response_guarantee, mara_number, bio, avatar_color
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (
            data['name'],
            data['firm'],
            data.get('tier', 'mid'),
            data['practiceArea'],
            json.dumps(data.get('specialties', [])),
            data['experienceYears'],
            data.get('caseCount', 0),
            data.get('successRate', 75),
            data['hourlyRateMin'],
            data['hourlyRateMax'],
            data['locationCity'],
            data['locationState'],
            1 if data.get('verified', False) else 0,
            1 if data.get('mediationCertified', False) else 0,
            1 if data.get('responseGuarantee', False) else 0,
            data.get('maraNumber'),
            data.get('bio'),
            data.get('avatarColor', '#000000')
        ))
        
        conn.commit()
        conn.close()
        
        return jsonify({'message': 'Lawyer created successfully'}), 201
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@lawyers_bp.route('/<int:lawyer_id>', methods=['PUT'])
@authenticate_token
@require_admin
def update_lawyer(lawyer_id):
    try:
        data = request.get_json()
        
        conn = get_db()
        cursor = conn.cursor()
        
        # Build update query dynamically
        updates = []
        values = []
        
        field_mapping = {
            'name': 'name',
            'firm': 'firm',
            'tier': 'tier',
            'practiceArea': 'practice_area',
            'specialties': 'specialties',
            'experienceYears': 'experience_years',
            'caseCount': 'case_count',
            'successRate': 'success_rate',
            'hourlyRateMin': 'hourly_rate_min',
            'hourlyRateMax': 'hourly_rate_max',
            'locationCity': 'location_city',
            'locationState': 'location_state',
            'verified': 'verified',
            'mediationCertified': 'mediation_certified',
            'responseGuarantee': 'response_guarantee',
            'maraNumber': 'mara_number',
            'bio': 'bio',
            'avatarColor': 'avatar_color'
        }
        
        for key, value in data.items():
            db_key = field_mapping.get(key)
            if db_key:
                if key == 'specialties' and isinstance(value, list):
                    updates.append(f'{db_key} = ?')
                    values.append(json.dumps(value))
                elif key in ['verified', 'mediationCertified', 'responseGuarantee']:
                    updates.append(f'{db_key} = ?')
                    values.append(1 if value else 0)
                else:
                    updates.append(f'{db_key} = ?')
                    values.append(value)
        
        if not updates:
            conn.close()
            return jsonify({'error': 'No valid fields to update'}), 400
        
        values.append(lawyer_id)
        query = f'UPDATE lawyers SET {", ".join(updates)} WHERE id = ?'
        
        cursor.execute(query, values)
        conn.commit()
        conn.close()
        
        return jsonify({'message': 'Lawyer updated successfully'})
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

@lawyers_bp.route('/<int:lawyer_id>', methods=['DELETE'])
@authenticate_token
@require_admin
def delete_lawyer(lawyer_id):
    try:
        conn = get_db()
        cursor = conn.cursor()
        cursor.execute('DELETE FROM lawyers WHERE id = ?', (lawyer_id,))
        conn.commit()
        conn.close()
        
        if cursor.rowcount == 0:
            return jsonify({'error': 'Lawyer not found'}), 404
        
        return jsonify({'message': 'Lawyer deleted successfully'})
        
    except Exception as e:
        return jsonify({'error': 'Internal server error'}), 500

File: backend/routes/lawyers.js
const express = require('express');
const router = express.Router();
const { getAll, getOne, runQuery } = require('../database/db');
const { authenticateToken, requireAdmin } = require('../middleware/auth');

// Get all lawyers with optional filters
router.get('/', async (req, res) => {
    try {
        const {
            practiceArea,
            state,
            minExperience,
            maxRate,
            responseGuarantee,
            sortBy = 'id'
        } = req.query;

        let query = 'SELECT * FROM lawyers WHERE 1=1';
        const params = [];

        if (practiceArea) {
            query += ' AND practice_area = ?';
            params.push(practiceArea);
        }

        if (state) {
            query += ' AND location_state = ?';
            params.push(state);
        }

        if (minExperience) {
            query += ' AND experience_years >= ?';
            params.push(parseInt(minExperience));
        }

        if (maxRate) {
            query += ' AND hourly_rate_min <= ?';
            params.push(parseFloat(maxRate));
        }

        if (responseGuarantee === 'true') {
            query += ' AND response_guarantee = 1';
        }

        // Sorting
        const validSorts = ['id', 'experience_years', 'hourly_rate_min', 'success_rate'];
        const sortField = validSorts.includes(sortBy) ? sortBy : 'id';
        query += ` ORDER BY ${sortField} DESC`;

        const lawyers = await getAll(query, params);

        // Parse specialties JSON
        const formattedLawyers = lawyers.map(lawyer => ({
            ...lawyer,
            specialties: lawyer.specialties ? JSON.parse(lawyer.specialties) : [],
            verified: Boolean(lawyer.verified),
            mediationCertified: Boolean(lawyer.mediation_certified),
            responseGuarantee: Boolean(lawyer.response_guarantee),
            practiceArea: lawyer.practice_area,
            experienceYears: lawyer.experience_years,
            caseCount: lawyer.case_count,
            successRate: lawyer.success_rate,
            hourlyRateMin: lawyer.hourly_rate_min,
            hourlyRateMax: lawyer.hourly_rate_max,
            locationCity: lawyer.location_city,
            locationState: lawyer.location_state,
            maraNumber: lawyer.mara_number,
            avatarColor: lawyer.avatar_color
        }));

        res.json({ lawyers: formattedLawyers });
    } catch (error) {
        console.error('Get lawyers error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Get single lawyer by ID
router.get('/:id', async (req, res) => {
    try {
        const lawyer = await getOne('SELECT * FROM lawyers WHERE id = ?', [req.params.id]);

        if (!lawyer) {
            return res.status(404).json({ error: 'Lawyer not found' });
        }

        const formattedLawyer = {
            ...lawyer,
            specialties: lawyer.specialties ? JSON.parse(lawyer.specialties) : [],
            verified: Boolean(lawyer.verified),
            mediationCertified: Boolean(lawyer.mediation_certified),
            responseGuarantee: Boolean(lawyer.response_guarantee),
            practiceArea: lawyer.practice_area,
            experienceYears: lawyer.experience_years,
            caseCount: lawyer.case_count,
            successRate: lawyer.success_rate,
            hourlyRateMin: lawyer.hourly_rate_min,
            hourlyRateMax: lawyer.hourly_rate_max,
            locationCity: lawyer.location_city,
            locationState: lawyer.location_state,
            maraNumber: lawyer.mara_number,
            avatarColor: lawyer.avatar_color
        };

        res.json({ lawyer: formattedLawyer });
    } catch (error) {
        console.error('Get lawyer error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Create new lawyer (admin only)
router.post('/', authenticateToken, requireAdmin, async (req, res) => {
    try {
        const {
            name,
            firm,
            tier = 'mid',
            practiceArea,
            specialties = [],
            experienceYears,
            caseCount = 0,
            successRate = 75,
            hourlyRateMin,
            hourlyRateMax,
            locationCity,
            locationState,
            verified = false,
            mediationCertified = false,
            responseGuarantee = false,
            maraNumber,
            bio,
            avatarColor
        } = req.body;

        if (!name || !firm || !practiceArea || !experienceYears || !hourlyRateMin || !hourlyRateMax || !locationCity || !locationState) {
            return res.status(400).json({ error: 'Missing required fields' });
        }

        const result = await runQuery(
            `INSERT INTO lawyers (
                name, firm, tier, practice_area, specialties, experience_years,
                case_count, success_rate, hourly_rate_min, hourly_rate_max,
                location_city, location_state, verified, mediation_certified,
                response_guarantee, mara_number, bio, avatar_color
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
            [
                name, firm, tier, practiceArea, JSON.stringify(specialties),
                experienceYears, caseCount, successRate, hourlyRateMin, hourlyRateMax,
                locationCity, locationState, verified ? 1 : 0, mediationCertified ? 1 : 0,
                responseGuarantee ? 1 : 0, maraNumber || null, bio || null,
                avatarColor || '#' + Math.floor(Math.random() * 16777215).toString(16)
            ]
        );

        const lawyer = await getOne('SELECT * FROM lawyers WHERE id = ?', [result.lastID]);

        res.status(201).json({
            message: 'Lawyer created successfully',
            lawyer: {
                ...lawyer,
                specialties: lawyer.specialties ? JSON.parse(lawyer.specialties) : []
            }
        });
    } catch (error) {
        console.error('Create lawyer error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Update lawyer (admin only)
router.put('/:id', authenticateToken, requireAdmin, async (req, res) => {
    try {
        const lawyerId = req.params.id;
        const updates = req.body;

        // Build dynamic update query
        const allowedFields = [
            'name', 'firm', 'tier', 'practice_area', 'specialties', 'experience_years',
            'case_count', 'success_rate', 'hourly_rate_min', 'hourly_rate_max',
            'location_city', 'location_state', 'verified', 'mediation_certified',
            'response_guarantee', 'mara_number', 'bio', 'avatar_color'
        ];

        const updateFields = [];
        const values = [];

        for (const [key, value] of Object.entries(updates)) {
            const dbKey = key.replace(/([A-Z])/g, '_$1').toLowerCase();
            if (allowedFields.includes(dbKey)) {
                if (key === 'specialties' && Array.isArray(value)) {
                    updateFields.push(`${dbKey} = ?`);
                    values.push(JSON.stringify(value));
                } else if (key === 'verified' || key === 'mediationCertified' || key === 'responseGuarantee') {
                    updateFields.push(`${dbKey} = ?`);
                    values.push(value ? 1 : 0);
                } else {
                    updateFields.push(`${dbKey} = ?`);
                    values.push(value);
                }
            }
        }

        if (updateFields.length === 0) {
            return res.status(400).json({ error: 'No valid fields to update' });
        }

        values.push(lawyerId);
        const query = `UPDATE lawyers SET ${updateFields.join(', ')} WHERE id = ?`;

        await runQuery(query, values);

        const updatedLawyer = await getOne('SELECT * FROM lawyers WHERE id = ?', [lawyerId]);

        res.json({
            message: 'Lawyer updated successfully',
            lawyer: updatedLawyer
        });
    } catch (error) {
        console.error('Update lawyer error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Delete lawyer (admin only)
router.delete('/:id', authenticateToken, requireAdmin, async (req, res) => {
    try {
        const result = await runQuery('DELETE FROM lawyers WHERE id = ?', [req.params.id]);

        if (result.changes === 0) {
            return res.status(404).json({ error: 'Lawyer not found' });
        }

        res.json({ message: 'Lawyer deleted successfully' });
    } catch (error) {
        console.error('Delete lawyer error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

module.exports = router;

